src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -205,27 +205,67 @@ static void VerifyDamOnDerivedAndBaseMethodsMatch (SymbolAnalysisContext context
 		{
-			if (FlowAnnotations.GetMethodReturnValueAnnotation (method) != FlowAnnotations.GetMethodReturnValueAnnotation (overriddenMethod))
+			var methodReturnAnnotations = FlowAnnotations.GetMethodReturnValueAnnotation (method);
+			var overriddenMethodReturnAnnotations = FlowAnnotations.GetMethodReturnValueAnnotation (overriddenMethod);
+			if (methodReturnAnnotations != overriddenMethodReturnAnnotations) {
+
+				(IMethodSymbol attributableMethod, DynamicallyAccessedMemberTypes missingAttribute) = GetTargetAndRequirements (method,
+					overriddenMethod, methodReturnAnnotations, overriddenMethodReturnAnnotations);
+
+				Location attributableSymbolLocation = attributableMethod.Locations[0];
+
+				(Location[]? sourceLocation, Dictionary<string, string?>? DAMArgs) = (!attributableSymbolLocation.IsInSource
+					|| (method.TryGetReturnAttribute (DynamicallyAccessedMembersAnalyzer.DynamicallyAccessedMembersAttribute, out var _)
+						&& overriddenMethod.TryGetReturnAttribute (DynamicallyAccessedMembersAnalyzer.DynamicallyAccessedMembersAttribute, out var _))
+						) ? (null, null) : CreateArguments (attributableSymbolLocation, missingAttribute);
+
 				context.ReportDiagnostic (Diagnostic.Create (
 					DiagnosticDescriptors.GetDiagnosticDescriptor (DiagnosticId.DynamicallyAccessedMembersMismatchOnMethodReturnValueBetweenOverrides),
-					method.Locations[0], method.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+					method.Locations[0], sourceLocation, DAMArgs?.ToImmutableDictionary (), method.GetDisplayName (), overriddenMethod.GetDisplayName ()));
+			}
 
 			for (int i = 0; i < method.Parameters.Length; i++) {
-				if (FlowAnnotations.GetMethodParameterAnnotation (method.Parameters[i]) != FlowAnnotations.GetMethodParameterAnnotation (overriddenMethod.Parameters[i])) {
+				var methodParameterAnnotation = FlowAnnotations.GetMethodParameterAnnotation (method.Parameters[i]);
+				var overriddenParameterAnnotation = FlowAnnotations.GetMethodParameterAnnotation (overriddenMethod.Parameters[i]);
+				if (methodParameterAnnotation != overriddenParameterAnnotation) {
+
+					(IMethodSymbol attributableMethod, DynamicallyAccessedMemberTypes missingAttribute) = GetTargetAndRequirements (method,
+						overriddenMethod, methodParameterAnnotation, overriddenParameterAnnotation);
+
+					Location attributableSymbolLocation = attributableMethod.Parameters[i].Locations[0];
+
+					(Location[]? sourceLocation, Dictionary<string, string?>? DAMArgs) = (!attributableSymbolLocation.IsInSource
+						|| (method.Parameters[i].TryGetAttribute (DynamicallyAccessedMembersAnalyzer.DynamicallyAccessedMembersAttribute, out var _)
+							&& overriddenMethod.Parameters[i].TryGetAttribute (DynamicallyAccessedMembersAnalyzer.DynamicallyAccessedMembersAttribute, out var _))
+							) ? (null, null) : CreateArguments (attributableSymbolLocation, missingAttribute);
```

Comment:
Looks like this block is repeated in each loop with only slightly different parameter values (`Parameters[i]` vs `TypeParameters[i]`). I think it would be worthwhile to refactor this to a local function with some comments for each case instead of having it uncommented and duplicated.

In fact, it looks like the whole loop can likely be refactored into a helper function, but that might be a little more difficult.

Reply:
The reason I left these in each loop as opposed to putting them in a helper is because for methods we use TryGetReturnAttribute, but for the two parameter loops we use TryGetAttribute. 

There are slight differences between each of the loops, so I decided it would be clunky to try to extract the similar-ish looking code, especially as there are only three cases. I did make two helper functions to reduce the amount of code in each block. To refactor everything we would have to feed in many arguments to a method and provide additional checks to see if we were working with a method, parameter, or typeparameter for both the method and the overridden method, which seems unneccessary given we already do these checks in the loop

Reply:
Yeah it looks like they might differ enough to not be worth de-duplicating. Adding some comments or splitting the condition into a few separate statements would make it more readable though.

We can look at ways to refactor this in the future. Refactoring this shouldn't hold up this PR.